evm_loader/cli/src/event_listener/listener_tracer.rs

```diff
@@ -7,27 +6,23 @@ use crate::{
 };
 
 pub trait ListenerTracer {
-    fn step (&mut self, mes: &StepTrace);
+    fn begin_step(&mut self, stack: Vec<[u8; 32]>, memory: Vec<u8>);
```

Comment:
There are two ways for the type system to enforce we always call an action that needs to begin and end, instead of having `begin` and `end` methods:

We can have a `step` method that accepts a closure. It calls the logic of `begin`, calls the closure, and then calls the `end` algorithm.

The other way is making `begin_step` create a struct that implements `Drop` and calling the logic of `end_step` in the `Drop`. This one may be more complex to handle.

But in both cases, we enforce that we always begin and end the step and never have a situation where a developer calls `begin` and forgets to call `end`.

Reply:
This doesn't apply here.
Imagine an event loop, where events arrives from somewhere outside of your control.
You have Begin Event and End Event. You always have an equal number of Begins and Ends.
But multiple Begin Events can appear in sequence and multiple End Events afterwards.